File: BoxNeighborCollection.cs

/*****************************************************************************
 *                                                                           *
 *  BoxNeighborCollection.cs                                                 *
 *  19 March 2007                                                            *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2007 Metaphysics Industries, Inc.                            *
 *                                                                           *
 *  Converted from C++ to C# on 10 November 2007                             *
 *                                                                           *
 *  A default implementation of a collection of boxes that automatically     *
 *    keeps changes to boxes' neighbors current.                             *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Collections;

namespace MetaphysicsIndustries.Crystalline
{
    [Serializable]
	public class BoxNeighborCollection : ICollection<Box>
	{
        private BoxNeighborCollection() { }
		public BoxNeighborCollection(Box parent, BoxOrientation orientation)
		{
			_parent = parent;
			_orientation = orientation;
			_set = new Set<Box>();
		}

		public virtual void Dispose()
		{
			this.Clear();
		}

		public virtual void Add(Box b)
		{
			if (b == null) { throw new ArgumentNullException("b"); }
			
			
			
			
			if (b != _parent && !this.Contains(b))
			{
				if (Orientation == BoxOrientation.Left ||
					Orientation == BoxOrientation.Right ||
					Orientation == BoxOrientation.Up ||
					Orientation == BoxOrientation.Down)
				{
					_set.Add(b);
				}
				else
				{
					throw new InvalidOperationException("unknown BoxOrientation");
				}
				if (Orientation == BoxOrientation.Left)
				{
					b.RightNeighbors.Add(_parent);
				}
				else if (Orientation == BoxOrientation.Right)
				{
					b.LeftNeighbors.Add(_parent);
				}
				else if (Orientation == BoxOrientation.Up)
				{
					b.DownNeighbors.Add(_parent);
				}
				else if (Orientation == BoxOrientation.Down)
				{
					b.UpNeighbors.Add(_parent);
				}
			}
		}

        public virtual void AddRange(IEnumerable<Box> items)
        {
            foreach (Box box in items)
            {
                Add(box);
            }
        }

		public virtual void Clear()
		{
			Box[]	r;
			r = new Box[Count];
			this.CopyTo(r, 0);
			foreach (Box b in r)
			{
				this.Remove(b);
			}
			_set.Clear();
		}

		public virtual bool Contains(Box b)
		{
			return _set.Contains(b);
		}

		public virtual bool Remove(Box b)
		{
			if (b == null) { throw new ArgumentNullException("b"); }
			if (this.Contains(b))
			{
				if (Orientation == BoxOrientation.Left ||
					Orientation == BoxOrientation.Right ||
					Orientation == BoxOrientation.Up ||
					Orientation == BoxOrientation.Down)
				{
					if (!_set.Remove(b)) { return false; }
				}
				else
				{
					throw new InvalidOperationException("unknown BoxOrientation");
				}
				if (Orientation == BoxOrientation.Left)
				{
					b.RightNeighbors.Remove(_parent);
				}
				else if (Orientation == BoxOrientation.Right)
				{
					b.LeftNeighbors.Remove(_parent);
				}
				else if (Orientation == BoxOrientation.Up)
				{
					b.DownNeighbors.Remove(_parent);
				}
				else if (Orientation == BoxOrientation.Down)
				{
					b.UpNeighbors.Remove(_parent);
				}
				return true;
			}
			return false;
		}

		public virtual void CopyTo(Box[] r, int i)
		{
			_set.CopyTo(r, i);
		}

		public virtual IEnumerator<Box> GetEnumerator()
		{
			return _set.GetEnumerator();
		}

		public virtual BoxOrientation Orientation
		{
			get
			{
				return _orientation;
			}
		}

		public virtual bool IsReadOnly
		{
			get
			{
				return _set.IsReadOnly;
			}
		}

		public virtual int Count
		{
			get
			{
				return _set.Count;
			}
		}

		  System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
		{
            return GetEnumerator();
		}

        private Box _parent;
        private BoxOrientation _orientation;
        private Set<Box> _set;
	}
}

File: Crystalline/BoxNeighborCollection.cpp

/*****************************************************************************
 *                                                                           *
 *  BoxNeighborCollection.cpp                                                        *
 *  19 March 2007                                                            *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2007 Metaphysics Industries                                  *
 *                                                                           *
 *  A default implementation of a collection of boxes that automatically     *
 *    keeps changes to boxes' neighbors current.                             *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "BoxNeighborCollection.h"



namespace MetaphysicsIndustries { namespace Crystalline { 
;



BoxNeighborCollection::BoxNeighborCollection(IBox^ __parent, BoxOrientation __orientation)
{
	this->_parent = __parent;
	this->_orientation = __orientation;
	this->_set = gcnew Set<IBox^>();
}

BoxNeighborCollection::~BoxNeighborCollection(void)
{
	this->Clear();

	this->_set->Clear();
	delete this->_set;
	this->_set = nullptr;
}

void BoxNeighborCollection::Add(IBox^ b)
{
	if (!b) { throw gcnew ArgumentNullException(__WCODESIG__); }

	//if (b == this->_parent)
	//{
	//	Debug::Fail("Adding a box to its own list of neighbors." + __WCODESIG__);
	//}

	if (b != this->_parent && !this->Contains(b))
	{
		if (this->Orientation == BoxOrientation::Left || 
			this->Orientation == BoxOrientation::Right || 
			this->Orientation == BoxOrientation::Up || 
			this->Orientation == BoxOrientation::Down)
		{
			this->_set->Add(b);
		}
		else
		{
			throw gcnew InvalidOperationException("unknown BoxOrientation" + __WCODESIG__);
		}

		if (this->Orientation == BoxOrientation::Left)
		{
			b->RightNeighbors->Add(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Right)
		{
			b->LeftNeighbors->Add(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Up)
		{
			b->DownNeighbors->Add(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Down)
		{
			b->UpNeighbors->Add(this->_parent);
		}
	}
}

bool BoxNeighborCollection::Remove(IBox^ b)
{
	if (!b) { throw gcnew ArgumentNullException(__WCODESIG__); }

	if (this->Contains(b))
	{
		if (this->Orientation == BoxOrientation::Left || 
			this->Orientation == BoxOrientation::Right || 
			this->Orientation == BoxOrientation::Up || 
			this->Orientation == BoxOrientation::Down)
		{
			if (!this->_set->Remove(b)) { return false; }
		}
		else
		{
			throw gcnew InvalidOperationException("unknown BoxOrientation" + __WCODESIG__);
		}

		if (this->Orientation == BoxOrientation::Left)
		{
			b->RightNeighbors->Remove(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Right)
		{
			b->LeftNeighbors->Remove(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Up)
		{
			b->DownNeighbors->Remove(this->_parent);
		}
		else if (this->Orientation == BoxOrientation::Down)
		{
			b->UpNeighbors->Remove(this->_parent);
		}

		return true;
	}

	return false;
}

bool BoxNeighborCollection::Contains(IBox^ b)
{
	return this->_set->Contains(b);
}

void BoxNeighborCollection::Clear(void)
{
	array<IBox^>^	r;
	r = gcnew array<IBox^>(this->Count);
	this->CopyTo(r, 0);
	for each (IBox^ b in r)
	{
		this->Remove(b);
	}
	this->_set->Clear();
}

void BoxNeighborCollection::CopyTo(array<IBox^>^ r, int i)
{
	this->_set->CopyTo(r, i);
}

IEnumerator<IBox^>^ BoxNeighborCollection::GetEnumerator(void)
{
	return this->_set->GetEnumerator();
}

int BoxNeighborCollection::Count::get(void)
{
	return this->_set->Count;
}

bool BoxNeighborCollection::IsReadOnly::get(void)
{
	return this->_set->IsReadOnly;
}

BoxOrientation BoxNeighborCollection::Orientation::get(void)
{
	return this->_orientation;
}

System::Collections::IEnumerator^ BoxNeighborCollection::GetEnumerator2(void)
{
	return this->GetEnumerator();
}






} } 


